# Python Data, 2025

---

* Úvod do NUMPY,
* základy NUMPY,
    - atributy,
    - metody,
    - spojování,
    - rozdělování,
    * cvičení 1,
    * cvičení 2,
* univerzální funkce, ~ufuncs,
    - úvod
    - typy funkcí
    * cvičení 3.

## Úvod do NUMPY

---

<img src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Ftse1.mm.bing.net%2Fth%3Fid%3DOIP.QQOtOJPya2YEO6IkVBnYjQHaDV%26pid%3DApi&f=1&ipt=245c27df971ff24527ca317988f6feda8b14851ef0a889fda7d5a10d114dc5e6&ipo=images" width="300" style="margin-left:auto; margin-right:auto"/>

### Proč umět NUMPY

---

Je to univerzální knihovna pro práci **s numerickými daty** v jazyce Python.

Funkcionalita této knihovny je využívána (skrze API) v jiných knihovnách jako *Pandas*, *SciPy*, *Matplotlib*, *scikit-learn*, *scikit-image*.

<br>

Nabízí efektivní nástroje pro práci **s většími datovými strukturami**.

Usnadňuje práci se **vícerozměrnými poli** a **maticovými datovými strukturami**.

### Kde zjistit více informací

---

Dokumentaci najdeš [zde](https://numpy.org/doc/stable/index.html#).

Zdrojový kód najdeš [zde](https://github.com/numpy/numpy).

### Jak jej nainstaluji

---

Instalace knihovny obecně (doporučeno v rámci virtuálního prostředí):

In [101]:
!pip install numpy

ERROR: Operation cancelled by user
^C
Traceback (most recent call last):
  File "/home/viegorova/.pyenv/versions/3.11.1/bin/pip", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/viegorova/.pyenv/versions/3.11.1/lib/python3.11/site-packages/pip/_internal/cli/main.py", line 70, in main
    return command.main(cmd_args)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/viegorova/.pyenv/versions/3.11.1/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 101, in main
    return self._main(args)
           ^^^^^^^^^^^^^^^^
  File "/home/viegorova/.pyenv/versions/3.11.1/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 216, in _main
    self.handle_pip_version_check(options)
  File "/home/viegorova/.pyenv/versions/3.11.1/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 190, in handle_pip_version_check
    pip_self_version_check(session, options)
  File "/home/viegorova/.pyenv/versions/3.11.1/lib/python3.11/site

...nebo:

In [ ]:
!conda install numpy

/bin/bash: conda: command not found


<br>

Nahrávání knihovny:

In [ ]:
import numpy

In [ ]:
import numpy as np

In [ ]:
from numpy import array

<br>

Používaná verze knihovny:

In [ ]:
numpy.__version__

'2.2.3'

### Srovnání datových typů

---

Efektivní práce s daty si zakládá na tom, jak jsou data uložená.

Proto je zásadní tušit jak objekty zpracovává **klasický Python** a `numpy`.

In [ ]:
x: int = 10
x: float = 10.0
x: str = "deset"

*Dynamické typování* v Pythonu vypadá jako dobrý nápad.

Přináší jistou flexibilitu a jednodušší používání.

Také s sebou ovšem nese jisté povinnosti, které by obyčejná proměnná **neměla normálně zpracovávat**.

V ukázce neukládáš pouze hodnoty, ale další doprovodné informace:
* **počet referencí**, kvůli alokaci/dealokaci paměti,
* **typ kódovací sady**,
* **velikost** následujících dat.

#### LIST v Pythonu

---

In [ ]:
muj_list: list = [10, 10.0, "deset"]

In [ ]:
datove_typy_listu = [type(hodnota) for hodnota in muj_list]

In [ ]:
datove_typy_listu

[int, float, str]

<br>

Je jasné, že tento *komfort* sebou nese i svoji cenu.

Protože toho každá hodnota s sebou nese další doplňující informace (datový typ, kódování, ...).

I kdyby byly hodnoty stejného datové typu, přesto s sebou přenášejí řadu balastu.

Tomu se snaží knihovna `numpy` pomáhat.

In [ ]:
!pip install numpy


[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


In [ ]:
import numpy as np

In [ ]:
# Helper function for finding out the real size of an object

import sys
import numpy as np

def get_size(obj, seen=None):
    size = sys.getsizeof(obj)
    if seen is None:
        seen = set()
    obj_id = id(obj)
    if obj_id in seen:
        return 0
    # Important mark as seen *before* entering recursion to gracefully handle
    # self-referential objects
    seen.add(obj_id)
    if isinstance(obj, dict):
        size += sum([get_size(v, seen) for v in obj.values()])
        size += sum([get_size(k, seen) for k in obj.keys()])
    elif hasattr(obj, '__dict__'):
        size += get_size(obj.__dict__, seen)
    elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes, bytearray)):
        size += sum([get_size(i, seen) for i in obj])
    return size

In [ ]:
# Numpy gives us a class for working with matrices: numpy.ndarray

np.ndarray((9))

array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

In [ ]:
# Numpy uses C data types for the numbers it contains, i.e. we have finite 
# integers with predefined size (np.int32, np.in64) and with possibility
# of overflow (a very popular source of errors)

pole_cela_cisla = np.ndarray((10000), dtype=np.int32)  
print(pole_cela_cisla.dtype, pole_cela_cisla.itemsize, pole_cela_cisla.itemsize * pole_cela_cisla.size)

int32 4 40000


In [ ]:
# .nbytes gives us the size of the data in the numpy array, but real size
# is bigger due to the small overhead introduced by the ndarray
import sys 
pole_cela_cisla.nbytes, sys.getsizeof(pole_cela_cisla)

(40000, 40112)

In [ ]:
seznam_cela_cisla = pole_cela_cisla.tolist()
get_size(seznam_cela_cisla, set())

171900

![memory](img/memory.png)

*https://medium.com/analytics-vidhya/master-numpy-in-45-minutes-74b2460ecb00*

In [ ]:
a = pole_cela_cisla[0]
a, type(a)

(np.int32(1739227984), numpy.int32)

In [ ]:
# ...but with caution

int32arr = np.ndarray((1, 1), dtype=np.int32)
int32arr[0, 0] = 1e15
int32arr

OverflowError: Python integer 1000000000000000 out of bounds for int32

In [ ]:
# The maximum allowed values for the data types can be extracted with np.iinfo

np.iinfo(np.int32).max, a + np.iinfo(np.int32).max, type(a + np.iinfo(np.int32).max)

/tmp/ipykernel_11882/643218094.py:3: RuntimeWarning: overflow encountered in scalar add
  np.iinfo(np.int32).max, a + np.iinfo(np.int32).max, type(a + np.iinfo(np.int32).max)


(2147483647, np.int32(-2147475457), numpy.int32)

In [ ]:
# And once again here: an overflow can happen!
# It happens A LOT :)

np.iinfo(np.int64).max, a + np.iinfo(np.int64).max, type(a + np.iinfo(np.int64).max)

OverflowError: Python integer 9223372036854775807 out of bounds for int32

In [ ]:
# We can easily convert between Python data types and the Numpy data types

b = int(a)
b, type(b)

(8192, int)

![ndarray](img/ndarray.png)


![block](img/block.png)
https://stackoverflow.com/questions/53097952/how-to-understand-numpy-strides-for-layman

![array](img/array.png)

In [ ]:
ciselne_pole = np.array([1, 2, 3])
ciselne_pole

array([1, 2, 3])

In [ ]:
ciselne_pole = np.array(range(6))
ciselne_pole

array([0, 1, 2, 3, 4, 5])

In [ ]:
ciselne_pole.dtype

dtype('int64')

In [ ]:
type(ciselne_pole)

numpy.ndarray

<br>

Na rozdíl od datového typu`list` v jazyce Python, pole v `numpy` obsahují stejný datový typ.

Pokud tomu tak není, `numpy` provede *upcast* (představ si datovou konverzi):

In [ ]:
mix_desetinne_cele_hodnoty = np.array((1, 2, 3, 3.1416))

In [ ]:
mix_desetinne_cele_hodnoty

array([1.    , 2.    , 3.    , 3.1416])

In [ ]:
mix_desetinne_cele_hodnoty.dtype

dtype('float64')

<br>

Pokud potřebuješ dodržet konkrétní datový typ, použij explicitně *parametr* `dtype`:

In [ ]:
jen_cele = np.array((1, 2, 3, 3.1416), dtype="int64")

In [ ]:
jen_cele

array([1, 2, 3, 3])

<br>

Datové typy se snaží implicitně konvertovat. Pokud je to možné.

In [ ]:
jen_cele = np.array((1, 2, "tři", 3.1416), dtype="int64")  # "tři" --> 3

ValueError: invalid literal for int() with base 10: 'tři'

In [ ]:
int("tři")

ValueError: invalid literal for int() with base 10: 'tři'

In [ ]:
jen_cele

array([1, 2, 3, 3])

<br>

`numpy` nabízí obrovskou škálu rychlých a efektivních způsobů vytváření polí a manipulace s číselnými daty v nich.

Pracuje s homogenními sekvenčnímy objekty.

<br>

## Základy NUMPY

---

Protože je `numpy` postaven na jazyku C, přebírá také podobné datové typy.

Některé datové typy:
    
| Datový typ v `numpy` | Popis |
| :- | :- |
| `bool_` | boolean `True`, `False` |
| `int16` | celá čísla (bajty od -32768 do 32767) |
| `int32` | celá čísla (bajty od -2147483648 do 2147483647)) |
| `int64` | celá čísla (bajty od -9223372036854775808 to 9223372036854775807) |
| `float64` | desetinné čísla |
| `complex64` | komplexní čísla, reprezentovaná dvěma 64-bitovými dese. čísly |

<br>

Dále `numpy` poskytuje `array` objekty.

In [ ]:
muj_list = [1, 2, 3]                      # 1D
muj_2D_list = [[1, 2], [3, 4]]            # 2D pole == matice
muj_3D_list = [[[1, 2], [2, 3]], [[..]]]  # 3D

Ty slouží jako prostředky pro elegantní manipulaci s daty (znáš z `pandas`).

Základní operace pro `array`:
* **atributy**,
* **indexování**,
* **řezání**,
* **RESHAPE**,
* **JOIN** a **MERGE** --> *konkatenace*, *split*

### Vytvoření ARRAY

---

In [ ]:
import numpy as np

<br>

#### Pomocí built-in sekvencí

---

Můžeš jednoduše konvertovat built-in datové typy v Pythonu na `numpy` pole:

In [ ]:
muj_range: range = range(3)
muj_list: list = [3, 4, 5]
muj_tuple: tuple = (6, 7, 8)

In [ ]:
muj_list_arr = np.array(muj_list)
muj_tuple_arr = np.array(muj_tuple)
muj_range_arr = np.array(muj_range)

In [ ]:
muj_list_arr

array([3, 4, 5])

In [ ]:
type(muj_list_arr)

numpy.ndarray

<br>

Atribut `dtype` neví nic o datovém typu objektu, jen jakým objekty je naplněný:

In [ ]:
muj_list_arr.dtype

dtype('int64')

In [ ]:
isinstance(muj_list_arr, np.ndarray)

True

In [ ]:
type(muj_tuple_arr)

numpy.ndarray

In [ ]:
muj_list_arr.shape

(3,)

In [ ]:
muj_list_arr = np.array([
    [1, 2, 3],
    [2, 3, 4],
    [3, 4, 5]
])
muj_list_arr

array([[1, 2, 3],
       [2, 3, 4],
       [3, 4, 5]])

In [ ]:
muj_list_arr.shape

(3, 3)

![empty](img/empty.png)
![zeros](img/zeros.png)

![size](img/size.png)

<br>

#### Pomocí nul, funkce `zeros`

---

In [ ]:
jen_nuly = np.zeros([2, 5])  # Royal (rows) Crown (columns) cola 

In [ ]:
jen_nuly

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [ ]:
jen_nuly.dtype

dtype('float64')

<br>

Pro jiný datový typ je potřeba explicitně zadat hodnotu:

In [ ]:
jen_nuly = np.zeros([2, 5], dtype='int64')  # Royal (rows) Crown (columns) cola 

In [ ]:
jen_nuly

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [ ]:
jen_nuly.dtype

dtype('int64')

<br>

#### Pomocí jedniček, funkce `ones`

---

In [ ]:
jen_jednicky = np.ones(shape=(2, 3), dtype=int)

Definice nového pole je možná i s klíčovými argumenty (parametr `shape`).

In [ ]:
# np.ones?

In [ ]:
jen_jednicky.dtype

dtype('int64')

In [ ]:
jen_jednicky

array([[1, 1, 1],
       [1, 1, 1]])

<br>

#### Pomocí defaultní výplňové hodnoty, funkce `full`

---

In [ ]:
jen_pi = np.full(shape=(3, 5), fill_value=np.pi)  # math, math.pi

In [ ]:
jen_pi

array([[3.14159265, 3.14159265, 3.14159265, 3.14159265, 3.14159265],
       [3.14159265, 3.14159265, 3.14159265, 3.14159265, 3.14159265],
       [3.14159265, 3.14159265, 3.14159265, 3.14159265, 3.14159265]])

<br>

#### Pomocí rozsahů, funkce `arange`

---

Velmi intuitivně zadavaná datová struktura, podobně jako u Python `range`:

In [ ]:
interval_arr = np.arange(start=0, stop=10, step=1)

In [ ]:
interval_arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
type(interval_arr)

numpy.ndarray

In [ ]:
interval_arr.dtype

dtype('int64')

<br>

Narozdíl od standardní knihovny umí `numpy` arrange objekt tvořit intervaly i s pomocí desetinných čísel.

In [ ]:
interval_arr = np.arange(start=0, stop=1, step=0.1, dtype='float64')

In [ ]:
interval_arr

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

<br>

#### Pomocí náhodných celých čísel, funkce `random`

---

In [ ]:
# from random import randint

In [ ]:
random_hodnoty = np.random.randint(low=0, high=6, size=(2, 4))

In [ ]:
random_hodnoty

array([[1, 4, 2, 0],
       [1, 0, 2, 3]])

In [ ]:
type(random_hodnoty)

numpy.ndarray

In [ ]:
poradi = len(range(10))  # 10

<br>

Občas je vhodnější tvořit rozsahy dynamicky, než statickou (hárdkódovanou) hodnotou:

In [ ]:
random_hodnoty = np.random.randint(low=0, high=66, size=poradi)

In [ ]:
tuple(random_hodnoty)

(np.int64(19),
 np.int64(45),
 np.int64(3),
 np.int64(28),
 np.int64(54),
 np.int64(2),
 np.int64(11),
 np.int64(37),
 np.int64(63),
 np.int64(7))

In [ ]:
%%timeit 
# The special functions are often optimized for this task 
# (optimization is in C, we do not need those details; but if you can use special function - do it!)
pole_s_0 = np.zeros((100, 100))

2.97 μs ± 272 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [ ]:
%%timeit 
pole_s_0 = np.full((100, 100), 0)

6.06 μs ± 312 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [ ]:
%%timeit 
seznam_s_0 = [[0] * 100 for _ in range(100)]

46.9 μs ± 2.26 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [ ]:
%%timeit 
pole_z_range = np.arange(100)

860 ns ± 44.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [ ]:
%%timeit 
seznam_z_range = list(range(100))

1.04 μs ± 48.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


<br>

#### Jednodimenzionální pole

---

In [ ]:
m1 = np.random.randint(low=1, high=10, size=6)

In [ ]:
m1

array([5, 8, 9, 1, 1, 7])

<br>

Tedy prakticky jednoduché pole, které se na první pohled neliší tolik od `list`.

#### Dvoudimenzionální pole

---

In [ ]:
m2 = np.random.randint(10, size=(2, 3))

In [ ]:
m2

array([[6, 0, 9],
       [1, 9, 8]])

<br>

Definicí oznamuješ:
1. `10`, vyber náhodné celé čísla z intervalu `0-9`,
2. vytvoř **2 řádky**,
3. .. o **3 sloupcích**.

#### Třídimenzionální pole

---

In [ ]:
m3 = np.random.randint(10, size=(2, 3, 4))  # implicitní zadání "high"

In [ ]:
m3

array([[[5, 1, 1, 7],
        [2, 5, 6, 9],
        [1, 2, 7, 1]],

       [[3, 1, 6, 7],
        [0, 5, 5, 3],
        [0, 3, 1, 3]]])

<br>

Definicí oznamuješ:
1. `10`, vyber náhodné celé čísla z intervalu `0-9`,
2. vytvoř **2 matice**,
3. .. o **3 řádcích**,
4. .. a **4 sloupcích**.

<br>

### Atributy pro ARRAY

---



Pokud budeš potřebovat prozkoumat existující `array`, můžeš vyzkoušet tyto atributy a metody:
1. **Dimenze**,
2. **tvar**,
3. **velikost**,
4. **datové typy**,
5. **bajtová velikost**.

#### Dimenze matice

---

In [ ]:
m3.ndim

3

In [ ]:
m2.ndim

2

<br>

#### Tvar matice

---

In [ ]:
m3.shape

(2, 3, 4)

In [ ]:
m2.shape

(2, 3)

In [ ]:
m2

array([[6, 0, 9],
       [1, 9, 8]])

<br>

#### Velikost matice

---

In [ ]:
m3.size

24

<br>

#### Datový typ

---

In [ ]:
m3.dtype

dtype('int64')

<br>

#### Velikost v bajtech

---

In [ ]:
m3.nbytes

192

<br>

Jde o součin atributů **size** * **itemsize**.

<br>

### Indexování ARRAY

---



Postup vypadá podobně jako pro sekvenční datové typy v Pythonu.

Pomocí hranaté závorky a celého čísla, můžeš zpřístupnit konkrétní hodnotu/ hodnoty.

<br>

##### Jednodimenzionální

In [ ]:
m1

array([5, 8, 9, 1, 1, 7])

In [ ]:
m1[0]

np.int64(5)

In [ ]:
m1[1]

np.int64(8)

In [ ]:
m1[-1]

np.int64(7)

<br>

##### Dvoudimenzionální

In [ ]:
m2

array([[6, 0, 9],
       [1, 9, 8]])

In [ ]:
m2[0]

array([6, 0, 9])

In [ ]:
m2[0][0]

np.int64(6)

In [ ]:
m2[0, 0]

np.int64(6)

<br>

Pomocí indexů zpřístupníš hodnotu na konkrétní pozici a můžeš ji modifikovat:

In [ ]:
m2

array([[6, 0, 9],
       [1, 9, 8]])

In [ ]:
el_m2 = m2[0, 0]

In [ ]:
el_m2

np.int64(6)

In [ ]:
m2[0, 0] = 11

In [ ]:
m2

array([[11,  0,  9],
       [ 1,  9,  8]])

<br>

Dávej pozor na datové typy. Pokud budeš chtít přidat jiný datový typ:

In [ ]:
m2[0, 0] = 11.111

In [ ]:
m2

array([[11,  0,  9],
       [ 1,  9,  8]])

Výsledek bude na pozadí *truncatovaný*.

In [ ]:
m2[0, 0] = "jedenáct"

ValueError: invalid literal for int() with base 10: 'jedenáct'

<br>

### Slicing ARRAY

---



Podobně jako u zpřístupňování jednotlivých hodnot, můžeš tvořit *subarrays*.

Postup je opět velmi podobný pro *slicing* u sekvenčních hodnot.

Je doporučené udávat všechny argumenty, protože implicitní dosazování defaultních hodnot může mást uživatele.

In [ ]:
m2

array([[11,  0,  9],
       [ 1,  9,  8]])

In [ ]:
m2[0:1]

array([[11,  0,  9]])

In [ ]:
m2[:, :2]

array([[11,  0],
       [ 1,  9]])

In [ ]:
m2[1, :2]

array([1, 9])

<br>

Pro více dimenzí, dva řádky a jeden sloupec:

In [ ]:
dva_radky_jeden_sloupec = m2[:2, :1]

In [ ]:
dva_radky_jeden_sloupec

array([[11],
       [ 1]])

In [ ]:
jeden_radek_jeden_sloupec = m2[::2, :1]

In [ ]:
jeden_radek_jeden_sloupec

array([[11]])

<br>

Pokud pracuješ se *subarrays*, používáš pohledy originální hodnoty.

In [ ]:
dva_radky_jeden_sloupec

array([[11],
       [ 1]])

In [ ]:
dva_radky_jeden_sloupec[0, 0] = 13

In [ ]:
dva_radky_jeden_sloupec

array([[13],
       [ 1]])

In [ ]:
m2

array([[13,  0,  9],
       [ 1,  9,  8]])

<br>

Pokud potřebuješ přesto nachystat kopii a ponechat původní hodnoty nezměněné, vyzkoušej metodu `copy`:

In [ ]:
kopie_dva_radky_jeden_sloupec = m2[:2, :1].copy()

In [ ]:
kopie_dva_radky_jeden_sloupec

array([[13],
       [ 1]])

In [ ]:
kopie_dva_radky_jeden_sloupec[0, 0] = 22

In [ ]:
kopie_dva_radky_jeden_sloupec

array([[22],
       [ 1]])

In [ ]:
m2

array([[13,  0,  9],
       [ 1,  9,  8]])

<br>

### Přetváření ARRAY

---



Jestli budeš potřebovat změnit tvar pole nebo matice, vyzkoušej metodu `reshape`:

In [ ]:
m2

array([[13,  0,  9],
       [ 1,  9,  8]])

In [ ]:
m2[0, 0] = 11

In [ ]:
m2

array([[11,  0,  9],
       [ 1,  9,  8]])

In [ ]:
m2.reshape((3, 2))

array([[11,  7],
       [ 7,  9],
       [ 3,  0]])

In [ ]:
m2

array([[11,  0,  9],
       [ 1,  9,  8]])

In [ ]:
m2_reshaped = m2.reshape((3, 2))
m2_reshaped

array([[11,  0],
       [ 9,  1],
       [ 9,  8]])

In [ ]:
# Often numpy gives just a view, not a copy
m2_reshaped[0, 0] = -1
m2_reshaped

array([[-1,  0],
       [ 9,  1],
       [ 9,  8]])

In [ ]:
m2

array([[-1,  0,  9],
       [ 1,  9,  8]])

<br>

Vytvoříš si jednodimenzionální pole s pomocí funkce `arange`:

In [182]:
pole_1D = np.arange(1, 11)

In [183]:
pole_1D

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

<br>

Nachystám 2D pole, matici, pomocí metody `reshape`:

In [184]:
matice = pole_1D.reshape(2, 5)

In [185]:
matice

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10]])

In [186]:
# Pocet prvku u reshape musi zustat stejny!
pole_1D.shape

(10,)

In [187]:
pole_1D.reshape((3, 3))

ValueError: cannot reshape array of size 10 into shape (3,3)

In [ ]:
# naopak resize umi i padding!

pole = np.array([1, 2, 3, 4, 5, 6])
pole.resize((2, 4))  # in-place!

In [189]:
pole

array([[1, 2, 3, 4],
       [5, 6, 0, 0]])

In [191]:
pole = np.array([1, 2, 3, 4, 5, 6])
pole.resize((2, 2))
pole

array([[1, 2],
       [3, 4]])

In [192]:
pole.resize((3, 2))

ValueError: cannot resize an array that references or is referenced
by another array in this way.
Use the np.resize function or refcheck=False

In [ ]:
np.resize(pole, (1, 2))  # NOT in-place

array([[1, 2]])

<br>

Pomocí metody `reshape` můžeš také měnit **řádkový vektor za sloupcový**:

In [102]:
import numpy as np

In [103]:
radkovy_vektor = np.array((11, 22, 33, 44))

In [104]:
radkovy_vektor

array([11, 22, 33, 44])

In [105]:
radkovy_vektor.shape

(4,)

In [106]:
r =  radkovy_vektor[np.newaxis, :]

In [107]:
r

array([[11, 22, 33, 44]])

In [108]:
r.shape

(1, 4)

In [109]:
r.T

array([[11],
       [22],
       [33],
       [44]])

In [113]:
radkovy_vektor[np.newaxis, :]

array([[11, 22, 33, 44]])

In [114]:
radkovy_vektor[np.newaxis, :].shape

(1, 4)

In [118]:
x = np.expand_dims(radkovy_vektor, axis=0)

In [119]:
x

array([[11, 22, 33, 44]])

In [120]:
x.shape

(1, 4)

In [121]:
radkovy_vektor[:, np.newaxis]

array([[11],
       [22],
       [33],
       [44]])

In [122]:
radkovy_vektor[:, np.newaxis].shape

(4, 1)

In [123]:
x = np.expand_dims(radkovy_vektor, axis=1)

In [ ]:
# np.expand_dims?

In [124]:
x

array([[11],
       [22],
       [33],
       [44]])

In [125]:
x.shape

(4, 1)

#### Ukázka

---

Potřebuješ převést 1D pole na 2D matici, kde budeš mít 7 hodnot na řádek.

In [193]:
temp_data = np.random.uniform(15, 36, size=30*7)  # 1D

In [194]:
# temp_data

In [195]:
temp_data.shape

(210,)

In [196]:
temp_data.ndim

1

In [197]:
############ Po Ut St Ct Pa So Ne
#01. tyden
#02. tyden
#03.
#...

In [198]:
temp_data.reshape(2, 7)

ValueError: cannot reshape array of size 210 into shape (2,7)

In [199]:
temp_data.reshape(-1, 7)

array([[21.33484849, 19.20984673, 26.08675266, 15.37656628, 28.68494677,
        32.06512728, 34.19756207],
       [29.72488191, 31.2137493 , 18.34785994, 17.56086412, 17.80692252,
        16.76075236, 35.68895612],
       [31.57654918, 23.36284721, 25.70320566, 24.97246219, 29.39093738,
        33.63187069, 23.30115123],
       [17.5678882 , 27.16064281, 23.53647334, 29.30275677, 33.06329541,
        26.59754632, 32.74553312],
       [22.98412173, 23.96708029, 27.28683255, 24.16436482, 15.06597222,
        24.45441452, 23.14672938],
       [20.19309736, 23.69525923, 19.52277467, 32.18129852, 21.3860548 ,
        32.93468029, 18.75976992],
       [35.17166606, 29.97141042, 20.07192013, 22.97023996, 27.939749  ,
        21.08875855, 26.94648746],
       [27.64091631, 16.68356837, 23.07965512, 28.71238907, 21.28865645,
        33.90954465, 24.91088813],
       [19.22344683, 27.43372617, 19.93385803, 26.06557841, 25.87180657,
        15.06405425, 32.08047807],
       [21.1121754 , 28.1380

In [200]:
tydenni_temp_data = temp_data.reshape(-1, 7)

In [201]:
# tydenni_temp_data

In [202]:
tydenni_temp_data.shape

(30, 7)

In [203]:
tydenni_temp_data.ndim

2

![agg](agg.jpg)

<br>

### Spojování a rozdělování

---



Předchozí ukázky se vztahovaly pouze k jednomu poli, k jedné matici.

Teď uvidíš, jak můžeš kombinovat několik těchto typů objektu.

#### Spojování

Pomocí metody `concatenate` spojíš dvě pole nebo dvě matice nebo více objektů:

In [126]:
pole_1 = np.arange(1, 6)
pole_2 = np.array((6, 7, 8))

In [127]:
pole_1

array([1, 2, 3, 4, 5])

In [128]:
pole_2

array([6, 7, 8])

In [129]:
spojene_pole = np.concatenate(pole_1, pole_2)

TypeError: only integer scalar arrays can be converted to a scalar index

In [130]:
spojene_pole = np.concatenate([pole_1, pole_2])

In [131]:
spojene_pole

array([1, 2, 3, 4, 5, 6, 7, 8])

In [132]:
pole_2D_1 = np.random.randint(low=-15, high=15, size=(3, 2))

In [133]:
pole_2D_1

array([[-13, -14],
       [-11, -12],
       [ -6,   2]])

In [134]:
pole_2D_2 = np.random.randint(low=-15, high=15, size=(5, 2))

In [135]:
pole_2D_2

array([[-11,   0],
       [ 10, -15],
       [-15,   4],
       [  6,  -6],
       [ -4, -13]])

In [136]:
pole_2D_3 = np.random.randint(low=-15, high=15, size=(3, 2))

In [137]:
pole_2D_3

array([[ -7, -11],
       [ -2,  12],
       [ -3,   7]])

In [138]:
spojene_matice = np.concatenate([pole_2D_1, pole_2D_3], axis=1)

In [139]:
spojene_matice

array([[-13, -14,  -7, -11],
       [-11, -12,  -2,  12],
       [ -6,   2,  -3,   7]])

In [140]:
spojene_matice = np.concatenate([pole_2D_1, pole_2D_2], axis=0)

In [141]:
spojene_matice

array([[-13, -14],
       [-11, -12],
       [ -6,   2],
       [-11,   0],
       [ 10, -15],
       [-15,   4],
       [  6,  -6],
       [ -4, -13]])

<br>

Pokud nemáš pole nebo matice stejného rozměru, je lepší pracovat s funkcemi:
1. `vstack`, vertikální napojení,
2. `hstack`, horizontální napojení,

In [142]:
pole_1

array([1, 2, 3, 4, 5])

In [143]:
pole_2D_1

array([[-13, -14],
       [-11, -12],
       [ -6,   2]])

In [149]:
pole_1.shape, pole_2D_1.shape

((5,), (3, 2))

In [154]:
rozsirene_2D = np.vstack([pole_1, pole_2D_1])

ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 5 and the array at index 1 has size 2

In [152]:
pole_1_cut = pole_1[np.newaxis, :2]
pole_1_cut.shape

(1, 2)

In [155]:
rozsirene_2D = np.vstack([pole_1_cut, pole_2D_1])

In [156]:
rozsirene_2D.shape

(4, 2)

In [148]:
rozsirene_2D

array([[  1,   2],
       [-13, -14],
       [-11, -12],
       [ -6,   2]])

In [158]:
# 2D -> 3D
pole2d = np.array([[1, 2, 3], [4, 5, 6]])
rozsirene = np.expand_dims(pole2d, axis=2)

In [159]:
rozsirene.shape

(2, 3, 1)

In [161]:
pole2d[:, :, np.newaxis].shape

(2, 3, 1)

In [195]:
# stack along new dimension
obrazek_r = np.random.random((4, 4))
obrazek_g = np.random.random((4, 4))
obrazek_b = np.random.random((4, 4))
np.stack([obrazek_r, obrazek_g, obrazek_b]).shape

(3, 4, 4)

#### Rozdělování

Opakem spojování polí a matic je rozdělování, které je možné provést pomocí funkcí:
1. `split`,
2. `hsplit`,
3. `vsplit`

In [162]:
rada_arr = np.random.randint(low=100, high=200, size=(12))

In [163]:
rada_arr.size

12

In [164]:
rada_arr

array([177, 163, 193, 135, 118, 135, 185, 142, 198, 105, 129, 186])

In [165]:
r1, r2, r3 = np.split(rada_arr, [4, 8])

In [166]:
r1, r2, r3

(array([177, 163, 193, 135]),
 array([118, 135, 185, 142]),
 array([198, 105, 129, 186]))

Takže dělení probíhá pomocí dělících indexů.

Tedy máš vždy N indexů a N + 1 polí.

In [167]:
matice = rada_arr.reshape(3, 4)

In [168]:
matice

array([[177, 163, 193, 135],
       [118, 135, 185, 142],
       [198, 105, 129, 186]])

In [169]:
leve_sl, prave_sl = np.hsplit(matice, 2)

In [170]:
# np.hsplit?

In [171]:
leve_sl

array([[177, 163],
       [118, 135],
       [198, 105]])

In [172]:
prave_sl

array([[193, 135],
       [185, 142],
       [129, 186]])

In [173]:
matice.shape

(3, 4)

In [ ]:
# musi byt delitelne poctem splitu!
leve_sl, prave_sl = np.hsplit(matice, 3)

ValueError: array split does not result in an equal division

In [174]:
leve_sl, prave_sl = np.hsplit(matice, 2)

In [175]:
prvni_sl, druhy_sl, treti_sl, ctvrty_sl = np.hsplit(matice, 4)

In [176]:
prvni_sl, druhy_sl, treti_sl, ctvrty_sl

(array([[177],
        [118],
        [198]]),
 array([[163],
        [135],
        [105]]),
 array([[193],
        [185],
        [129]]),
 array([[135],
        [142],
        [186]]))

In [181]:
# non-equal splits again
split1, split2, split3 = np.split(matice, [2, 3], axis=1)
split1.shape, split2.shape, split3.shape

((3, 2), (3, 1), (3, 1))

<br>

##### **🧠 CVIČENÍ 🧠, procvič si základy v NUMPY**

In [ ]:
import numpy

In [ ]:
# 01. Použij NUMPY funkci, která vytvoří 1D pole od 5, do 50 po 5ti

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    np.arange(5, 51, 5)
    ```
</details>

In [ ]:
# 02. Jak vytvoříš stejné pole pomocí funkce RANGE a LIST?

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    np.array([cislo for cislo in range(5, 51, 5)])
    ```
</details>

In [ ]:
# 03. Jak můžeš vytvořit pole z kapitálek písmen od A do Z?

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    np.array([chr(index) for index in range(ord("A"), ord("Z") + 1)])
    ```
</details>

In [ ]:
# 04. Jak můžeš vytvořit 1D pole z 6ti náhodných čísel od 1 do 10 (včetně 10)?

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    np.random.randint(1, 10, size=6)
    ```
</details>

In [ ]:
# 05. Jak vytvoříš 2D matici 4 x 5 pole s náhodnými čísly od 1 do 50 (včetně 50)?

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    numpy.random.randint(1, 51, size=(4, 5))
    ```
</details>

<br>

##### **🧠 CVIČENÍ 🧠, procvič si základy v NUMPY**

Máš teplotní záznamy získané ze dvou různých meteorologických stanic.

Každá stanice poskytuje teplotní záznamy za posledních 10 dní, které jsou uloženy ve formátu seznamu. Tvým úkolem je:

In [ ]:
import numpy as np

In [ ]:
stanice_c1_teploty = [21.5, 22.3, 20.7, 19.8, 23.1, 22.6, 20.4, 21.8, 22.9, 21.2]
stanice_c2_teploty = [18.5, 19.1, 18.7, 17.8, 21.3, 20.6, 18.5, 19.7, 20.5, 19.2]

In [ ]:
# Převeď seznamy teplot na NumPy pole.

In [ ]:
# spoj oba záznamy do jednoho pole pomocí funkce 'concatenate'.

In [ ]:
# rozděl sloučené pole na dvě části: prvních 5 dní a druhých 5 dní.

In [ ]:
# vypočítej průměrnou teplotu pro prvních 5 dní a druhých 5 dní.

In [ ]:
# Převeď seznamy teplot na NumPy pole.
teploty1_np = np.array( stanice_c1_teploty )
teploty2_np = np.array( stanice_c2_teploty )

In [ ]:
# spoj oba záznamy do jednoho pole pomocí funkce 'concatenate'.
teploty = np.concatenate( [teploty1_np, teploty2_np] )

In [ ]:
# rozděl sloučené pole na dvě části: prvních 5 dní a druhých 5 dní.
daysN = 5
teploty_fix = np.concatenate( [teploty, np.full( ( daysN - teploty.size % daysN ) % daysN, np.nan )] )
teploty_po_N_dnech = np.reshape( teploty_fix, ( -1, daysN ) )

In [ ]:
# vypočítej průměrnou teplotu pro každou část
prumerne_teploty = teploty_po_N_dnech.mean( axis=1 ).round( 3 )
prumerne_teploty_s_popiskem = np.array( list( zip( [
    'Prvních 5 dní - první stanice',
    'Druhých 5 dní - první stanice',
    'Prvních 5 dní - druhá stanice',
    'Druhých 5 dní - druhá stanice',
], prumerne_teploty ) ) )
display( prumerne_teploty_s_popiskem )

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    stanice_c1_teploty = [21.5, 22.3, 20.7, 19.8, 23.1, 22.6, 20.4, 21.8, 22.9, 21.2]
    stanice_c2_teploty = [18.5, 19.1, 18.7, 17.8, 21.3, 20.6, 18.5, 19.7, 20.5, 19.2]
    
    stanice_c1_teploty_pole = np.array(stanice_c1_teploty)
    stanice_c2_teploty_pole = np.array(stanice_c2_teploty)
    
    spojene_teploty = np.concatenate((stanice_c1_teploty_pole, stanice_c2_teploty_pole))
    
    upraveny_tvar = spojene_teploty.reshape(2, 10)
    prvnich_pet_dni, dalsich_pet_dni = np.hsplit(upraveny_tvar, 5)
    
    prumer_prvnich_deset_dni = np.mean(prvnich_deset_dni)
    prumer_druhych_deset_dni = np.mean(dalsich_deset_dni)
    ```
</details>

## Matematické operace, ufuncs
---

In [241]:
pole = np.arange(100)
pole.shape

(100,)

In [242]:
pole + 1

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [243]:
pole * 2

array([  0,   2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,
        26,  28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,
        52,  54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,
        78,  80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102,
       104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128,
       130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154,
       156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180,
       182, 184, 186, 188, 190, 192, 194, 196, 198])

In [244]:
%%timeit 
pole + 1

1.49 μs ± 117 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [245]:
seznam = pole.tolist()

In [246]:
%%timeit
[hodnota + 1 for hodnota in seznam]

4.03 μs ± 480 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


<br>

### Univerzální funkce, UFUNCS

Výpočty na polích a maticích `numpy` mohou být **velmi rychlé** nebo **velmi pomalé**.

Klíčem k rychlé práci, je použití vektorizovaných operací, které jsou obvykle implementovány prostřednictvím *univerzálními funkcemi*.

Jde tedy o funkce, které pracují na objektech polí a matic, prvek po prvku.
Umožňují efektivní operace na polích, což vede k výrazně rychlejším výsledkům než pomocí běžných **Pythonových funkcí a cyklů**.

Ukázkou může být sčítání hodnot pro dvě dlouhé matice:

In [256]:
import numpy as np

In [257]:
pole_1 = np.random.randint(1, 100, size=10_000_000)

In [258]:
pole_2 = np.random.randint(100, 255, size=10_000_000)

In [259]:
pole_a = np.array([1, 2 ,3])
pole_b = np.array([4, 4, 4])

In [260]:
pole_a + pole_b

array([5, 6, 7])

<br>

Pro sečtení hodnot z obou polí můžeš pracovat s funkce `add`:

In [261]:
%timeit result_loop = [x + y for x, y in zip(pole_1, pole_2)]

2.28 s ± 145 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [262]:
%timeit secteno_ufuncs = np.add(pole_1, pole_2)

28.3 ms ± 1.95 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


<br>

Můžeš si všimnout, že řešení s výstupem v `secteno_ufuncs` je skoro 5~10x rychlejší.

<br>

Z důvodů použití dynamického, interpretovaného jazyka v rámci CPythonu ti jeho flexibilita neumožní kompilovat jednotlivé operace až na efektivní strojový kód (`01101101110101`).

Určitě existují derivace projektů jako *PyPy*, *Cython*, *Numba*.

### Úvod do UFUNCS

---

Pro mnoho typů operací poskytuje `numpy` praktické rozhraní právě pro tento typ operací,
staticky typované, zkompilované rutiny.

Jedná se o tzv. **vektorizované operace**.

Pro aplikace takových funkcí můžeš jednoduše vybrat konkrétní funkci a aplikovat ji na pole.

In [263]:
pole_3 = np.arange(10)

In [264]:
pole_4 = np.arange(11, 21)

In [265]:
pole_3, pole_4

(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
 array([11, 12, 13, 14, 15, 16, 17, 18, 19, 20]))

In [266]:
pole_3 / pole_4

array([0.        , 0.08333333, 0.15384615, 0.21428571, 0.26666667,
       0.3125    , 0.35294118, 0.38888889, 0.42105263, 0.45      ])

<br>

Počítání přes **ufuncs** je téměř pokaždé efektivnější než aplikace např. obyčejných smyček v Pythonu.

Kromě klasických operátorů, můžeš aplikovat samotné funkce:

In [267]:
np.divide(pole_3, pole_4)

array([0.        , 0.08333333, 0.15384615, 0.21428571, 0.26666667,
       0.3125    , 0.35294118, 0.38888889, 0.42105263, 0.45      ])

<br>

Výpis některých UFUNCS:

| Operátor | Jméno funkce | Popis |
| :- | :- | :- |
| `+` |  `np.add` |  Sčítání (př., `1 + 1 = 2` ) |
| `-` |  `np.subtract` | odčítání (př., `3 - 2 = 1` ) | 
| `-` | `np.negative` |  unární negace (př., `-2` ) | 
| `*` | `np.multiply` |  násobení (př., `2 * 3 = 6` ) |
| `/` | `np.divide` |  dělení (př., `3 / 2 = 1.5` ) |
| `//` |  `np.floor_divide` | celočíselné dělení (př., `3 // 2 = 1` ) |
| `**` | `np.power` | umocňování (př., `2 ** 3 = 8` ) | 
| `%` | `np.mod` | modulo (př., `9 % 4 = 1` ) |

<br>

Další kombinací můžeš pracovat třeba **s absolutními hodnotami**:

In [268]:
negativni_cisla = np.array((-5, -4, -3, -2, -1))

In [269]:
abs(negativni_cisla)

array([5, 4, 3, 2, 1])

In [270]:
type(abs)

builtin_function_or_method

<br>

Pomocí ufuncs použiješ funkci `absolute`, která je ekvivalentem v rámci `numpy`:

In [271]:
np.absolute(negativni_cisla)

array([5, 4, 3, 2, 1])

In [272]:
np.abs(negativni_cisla)

array([5, 4, 3, 2, 1])

<br>

Práce s **exponenty** a **umocňování**:

In [277]:
ciselne_hodnoty = np.arange(6)

In [278]:
ciselne_hodnoty

array([0, 1, 2, 3, 4, 5])

In [279]:
np.exp2(ciselne_hodnoty)      # 2^0, 2^1, 2^2, 2^3, ...

array([ 1.,  2.,  4.,  8., 16., 32.])

In [280]:
np.power(3, ciselne_hodnoty)  # 3^0, 3^1, ...

array([  1,   3,   9,  27,  81, 243])

<br>

Knihovna `numpy` potom nabízí další specializované funkce (logaritmické, trigonometrické, aj.)

In [281]:
A = np.random.random((10, 3))
B = np.random.random((3, 5))
A.shape, B.shape

((10, 3), (3, 5))

In [ ]:
%%timeit 
A @ B

In [ ]:
A_seznam = A.tolist() 
B_seznam = B.tolist()

In [ ]:
def matmul(A, B):
    return [[sum(A[i][k] * B[k][j] for k in range(len(B))) for j in range(len(B[0]))] for i in range(len(A))]


In [ ]:
%%timeit 
matmul(A_seznam, B_seznam)

In [ ]:
A = np.random.random((3, 3))
np.linalg.inv(A)

### Statistické funkce

In [63]:
hodnoty = [1, 2, 3, 4, 5, 6, 7]
np.mean(hodnoty)

np.float64(4.0)

In [65]:
np.std(hodnoty)

np.float64(2.0)

In [66]:
np.median(hodnoty)

np.float64(4.0)

### Funkce pro porovnání

In [67]:
x = np.array([1, 2, 3])
y = np.array([3, 2, 1])
np.greater(x, y)

array([False, False,  True])

In [68]:
x > y

array([False, False,  True])

In [69]:
np.less(x, y)

array([ True, False, False])

In [70]:
x < y

array([ True, False, False])

In [71]:
np.equal(x, y)

array([False,  True, False])

In [72]:
x == y

array([False,  True, False])

In [73]:
bool_pole = np.array([True, False, True])

In [74]:
np.logical_not(bool_pole)

array([False,  True, False])

In [75]:
np.logical_and(bool_pole, np.logical_not(bool_pole))

array([False, False, False])

In [76]:
np.logical_or(bool_pole, np.logical_not(bool_pole))

array([ True,  True,  True])

In [77]:
x[x < y]

array([1])

In [78]:
x[x == y]

array([2])

In [79]:
x[x <= y]

array([1, 2])

In [81]:
x[(x < y) & (x == y)]

array([], dtype=int64)

In [82]:
x[(x < y) | (x == y)]

array([1, 2])

In [84]:
x[(x < y) | ~(x < y)]

array([1, 2, 3])

### Vektorizované uživatelské funkce

In [ ]:
def danova_sazba(prijem):
    if prijem < 30000:
        return prijem * 0.1   # 10% daň
    elif prijem < 70000:
        return prijem * 0.2   # 20% daň
    else:
        return prijem * 0.3   # 30% daň

In [294]:
prijmy = np.random.randint(20000, 120000, size=1000)

In [295]:
vektorova_dan = np.vectorize(danova_sazba)

In [297]:
dane = vektorova_dan(prijmy)
dane[:5]

array([ 8505.2, 34369.5, 26481.9, 12203.4, 13546.4])

In [95]:
%%timeit 
# Bohuzel np.vectorize je spis pro pohodlny zapis, ne pro rychlost...
vektorova_dan(prijmy)

271 μs ± 13.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [96]:
prijmy_seznam = prijmy.tolist()

In [97]:
%%timeit 
for prijem in prijmy_seznam:
    vektorova_dan(prijem)

16.9 ms ± 986 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [98]:
%%timeit
[vektorova_dan(prijem) for prijem in prijmy_seznam]

18.5 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [99]:
def dan_numpy_where(prijmy_np):
    return np.where(prijmy_np < 30000, prijmy_np * 0.1,
                    np.where(prijmy_np < 70000, prijmy_np * 0.2, prijmy_np * 0.3))

In [100]:
%%timeit 
dan_numpy_where(prijmy)

22.6 μs ± 665 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


<br>

##### **🧠 CVIČENÍ 🧠, procvič si základy UFUNCS v NUMPY**

Máte prodejní data dvou obchodů za poslední 3 měsíce. Každý obchod poskytuje seznam svých prodejů za každý měsíc. Vaším úkolem je:

In [ ]:
obchod_1_prodej = [1000, 1200, 1300]
obchod_2_prodej = [900, 1100, 1500]

In [ ]:
# Převést seznamy prodejů na NumPy pole.
import numpy as np

In [ ]:
# 1. Sloučit prodejní data obou obchodů do jednoho pole pomocí funkce concatenate

In [ ]:
# 2. Rozdělit sloučené pole na dvě části, které odpovídají měsícům ..
# .. pod sebou u obou obchodů

In [ ]:
# 3. Vypočítat celkový prodej pro každý měsíc

In [ ]:
# 4. Vypočítat procentuální změnu mezi prvním a druhým měsícem ..
# .. a mezi druhým a třetím měsícem mezi celkovými měs. součty

<details>
    <summary>▶️ Řešení</summary>
    
    ```python
    obchod_1_prodej = [1000, 1200, 1300]
    obchod_2_prodej = [900, 1100, 1500]
    
    import numpy as np

    # 1. Sloučit prodejní data obou obchodů do jednoho pole pomocí funkce concatenate
    sloucene_prodeje = np.concatenate((obchod_1_prodej_pole, obchod_2_prodej_pole))
    
    # 2. Rozdělit sloučené pole na dvě části, které odpovídají měsícům ..
    # .. pod sebou u obou obchodů
    rozdelene_po_mesicich = sloucene_prodeje.reshape(2, -1)
    
    # 3. Vypočítat celkový prodej pro každý měsíc
    mesic_1, mesic_2, mesic_3 = np.hsplit(rozdelene_po_mesicich, 3)
    celkem_mesice = np.array((mesic_1.sum(), mesic_2.sum(), mesic_3.sum()))
    
    # 4. Vypočítat procentuální změnu mezi prvním a druhým měsícem ..
    # .. a mezi druhým a třetím měsícem mezi celkovými měs. součty
    rozdily = np.diff(celkem_mesice)
    np.divide(rozdily, celkem_mesice[:2]) * 100
    ```
</details>

---